src/wallet/wallet.cpp

```diff
@@ -100,8 +100,50 @@ libzcash::PaymentAddress CWallet::GenerateNewZKey()
     return addr;
 }
 
+// Generate a new Sapling spending key and return its public payment address
+SaplingPaymentAddress CWallet::GenerateNewSaplingZKey()
+{
+    AssertLockHeld(cs_wallet); // mapZKeyMetadata
+    
+    auto sk = SaplingSpendingKey::random();
+    auto fvk = sk.full_viewing_key();
+    auto addr = sk.default_address();
+    
+    // Check for collision, even though it is unlikely to ever occur
+    if (CCryptoKeyStore::HaveSaplingSpendingKey(fvk))
+    throw std::runtime_error("CWallet::GenerateNewSaplingZKey(): Collision detected");
+    
+    // Create new metadata
+    int64_t nCreationTime = GetTime();
+    mapSaplingZKeyMetadata[addr] = CKeyMetadata(nCreationTime);
+    
+    if (!AddSaplingZKey(sk)) {
+        throw std::runtime_error("CWallet::GenerateNewSaplingZKey(): AddSaplingZKey failed");
+    }
+    // return default sapling payment address.
+    return addr;
+}
+
+// Add spending key to keystore 
+bool CWallet::AddSaplingZKey(const libzcash::SaplingSpendingKey &sk)
+{
+    AssertLockHeld(cs_wallet); // mapSaplingZKeyMetadata
+
+    if (!CCryptoKeyStore::AddSaplingSpendingKey(sk)) {
```

Comment:
I suspect this call will fail, because this PR doesn't update `CCryptoKeyStore` with the new methods. Please add a test that calls `GenerateNewSaplingZKey()` and/or `AddSaplingZKey()` to ensure the pathway is exercised.

Reply:
If the ongoing `try` passes, I will downgrade this to non-blocking in the interest of unblocking downstream PRs.

Reply:
I updated `CCryptoKeyStore` to add sapling spending keys, and added a test calling  `GenerateNewSaplingZKey()` and `AddSaplingZKey()`, going to run a try again